samples/minimaloboe/src/main/java/com/example/minimaloboe/MainActivity.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.minimaloboe
+
+import android.os.Bundle
+import androidx.activity.ComponentActivity
+import androidx.activity.compose.setContent
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.material.*
+import androidx.compose.runtime.*
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.lifecycle.ViewModel
+import com.example.minimaloboe.ui.theme.SamplesTheme
+import kotlinx.coroutines.*
+
+class MainActivity : ComponentActivity() {
+
+    var mExampleViewModel = ExampleViewModel()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContent {
+            SamplesTheme {
+                // A surface container using the 'background' color from the theme
+                Surface(
+                    modifier = Modifier.fillMaxSize(),
+                    color = MaterialTheme.colors.background
+                ) {
+                    MainControls(mExampleViewModel)
+                }
+            }
+        }
+    }
+
+    override fun onStop() {
+        // call the superclass method first
+        super.onStop()
+        mExampleViewModel.stopAudio()
+    }
+}
+
+class ExampleViewModel : ViewModel() {
+
+    var audioPlayer = AudioPlayer()
+
+    fun startAudio(): Int {
+        return audioPlayer.startAudio()
+    }
+
+    fun stopAudio(): Int {
+        return audioPlayer.stopAudio()
+    }
+}
+
+@Composable
+fun MainControls(viewModel: ExampleViewModel = ExampleViewModel()) {
```

Comment:
1. viewModel shouldn't be instantiated like that, but you can do something very similiar with hilt viewmodel integration `viewModel: ExampleViewModel = viewModel()`. See [here](https://developer.android.com/jetpack/compose/libraries#hilt)
2. Composable should provide a modifier argument. See [here](https://chris.banes.dev/always-provide-a-modifier/)

Reply:
I will add a modifier argument.

Reply:
I think this obsolete with Don's changes.